libs/collab-stream/src/client.rs

```diff
@@ -66,29 +84,143 @@ impl CollabRedisStream {
     group.ensure_consumer_group().await?;
     Ok(group)
   }
+
+  pub fn collab_update_sink(&self, workspace_id: &str, object_id: &str) -> CollabUpdateSink {
+    let stream_key = CollabStreamUpdate::stream_key(workspace_id, object_id);
+    CollabUpdateSink::new(self.connection_manager.clone(), stream_key)
+  }
+
+  pub fn awareness_update_sink(&self, workspace_id: &str, object_id: &str) -> AwarenessUpdateSink {
+    let stream_key = AwarenessStreamUpdate::stream_key(workspace_id, object_id);
+    AwarenessUpdateSink::new(self.connection_manager.clone(), stream_key)
+  }
+
+  pub fn collab_updates(
+    &self,
+    workspace_id: &str,
+    object_id: &str,
+    since: Option<MessageId>,
+    live: bool,
+  ) -> impl Stream<Item = Result<(MessageId, CollabStreamUpdate), StreamError>> {
+    let mut conn = self.connection_manager.clone();
+    let stream_key = CollabStreamUpdate::stream_key(workspace_id, object_id);
+    let read_options = StreamReadOptions::default().count(100);
+    let mut since = since.unwrap_or_default();
+
+    // instead of doing active blocking xread, use an exponential backoff to wait asynchronously
+    // the longer Redis stream is inactive the longer the pause will become, and it will
+    // reset once we get new data onboard
+    let mut backoff = ExponentialBackoff::new(Duration::from_millis(100), Duration::from_secs(10));
+    async_stream::try_stream! {
+      let until = if live {
+        None
+      } else {
+        let mut reply: StreamRangeReply = conn.xrevrange_count(&stream_key, "+", "-", 1).await?;
+        match reply.ids.pop().map(|stream_id| stream_id.id) {
+          None => Some(MessageId::default()),
+          Some(id) => Some(MessageId::try_from(id)?),
+        }
+      };
+      while until.is_none() || since < until.unwrap() {
```

Comment:
it looks more safer by using
```rust
while until.map_or(true, |u| since < u) {}
```

Reply:
I made it other (arguably better) way that doesn't require pre-fetching final MessageId.